lib/pulselistener.js

```diff
@@ -79,56 +110,102 @@ util.inherits(PulseConnection, events.EventEmitter);
 
 /** Returns a promise for a connection */
 PulseConnection.prototype.connect = function() {
-  var that = this;
+  var self = this;
 
-  // Connection if we have one
-  if (this._conn) {
-    return Promise.resolve(this._conn);
+  // If connecting we're done
+  if (this._connecting) {
+    return this._connecting;
   }
 
-  // If currently connecting, give a promise for this result
-  var retval = new Promise(function(accept, reject) {
-    that.once('connection', function(err, conn) {
-      if (err) {
-        return reject(err);
+  // Function to attempt connection
+  var connect = function() {
+    return new Promise(function(accept) {
+      var attempts = self._reconnectAttempts;
+      if (attempts > 0) {
+        // Construct the usual delay and retry after sleeping
+        var delay = Math.pow(2, attempts) * self._options.delayFactor;
+        var rf = self._options.randomizationFactor;
+        delay = delay * (Math.random() * 2 * rf + 1 - rf);
+        delay = Math.min(delay, self._options.maxDelay);
+        return setTimeout(accept, delay);
```

Comment:
I see this pattern all around. It'd be better to use something reusable. James has written this wrapper at some point: https://github.com/lightsofapollo/promise-retries. Probably it can be extended to use randomness and backoffs.